lib/srv/git/key_manager.go

```diff
@@ -0,0 +1,160 @@
+func (c *KeyManagerConfig) CheckAndSetDefaults() error {
+	if c.ParentContext == nil {
+		return trace.BadParameter("missing parameter ParentContext")
+	}
+	if c.AuthClient == nil {
+		return trace.BadParameter("missing parameter AuthClient")
+	}
+	if c.AccessPoint == nil {
+		return trace.BadParameter("missing parameter AccessPoint")
+	}
+	if c.Logger == nil {
+		c.Logger = slog.With(teleport.ComponentKey, teleport.ComponentGit)
+	}
+	if c.githubServerKeys == nil {
+		c.githubServerKeys = newGitHubKeyDownloader()
+	}
+	return nil
+}
+
+// KeyManager manages and caches remote server keys.
+type KeyManager struct {
+	cfg *KeyManagerConfig
+}
+
+// NewKeyManager creates a service that manages and caches remote server keys.
+// TODO(greedy52) move user cert generation here with caching.
+func NewKeyManager(cfg *KeyManagerConfig) (*KeyManager, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	m := &KeyManager{
+		cfg: cfg,
+	}
+
+	if err := m.startWatcher(cfg.ParentContext); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return m, nil
+}
+
+func (m *KeyManager) startWatcher(ctx context.Context) error {
+	watcher, err := services.NewGitServerWatcher(ctx, services.GitServerWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component: teleport.ComponentGit,
+			Logger:    m.cfg.Logger,
+			Client:    m.cfg.AuthClient,
+		},
+		GitServerGetter:       m.cfg.AccessPoint.GitServerReadOnlyClient(),
+		EnableUpdateBroadcast: true,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Start background downloads only when git_servers are found.
+	// TODO(greedy52) use a reconciler and start downloader by type.
+	go func() {
+		defer m.cfg.Logger.DebugContext(ctx, "Git server resource watcher done.")
+		defer watcher.Close()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case resources := <-watcher.ResourcesC:
+				m.cfg.Logger.DebugContext(ctx, "Received git server resources from watcher", "len", len(resources))
+				if len(resources) > 0 {
+					go m.cfg.githubServerKeys.Start(ctx)
+					return
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+// HostKeyCallback creates an ssh.HostKeyCallback for verifying the target git-hosting service.
+func (m *KeyManager) HostKeyCallback(targetServer types.Server) ssh.HostKeyCallback {
+	return func(_ string, _ net.Addr, key ssh.PublicKey) error {
+		switch targetServer.GetSubKind() {
+		case types.SubKindGitHub:
+			return trace.Wrap(m.verifyGitHub(key))
+		default:
+			return trace.BadParameter("unsupported subkind %q", targetServer.GetSubKind())
+		}
+	}
+}
+
+func (m *KeyManager) verifyGitHub(key ssh.PublicKey) error {
+	knownKeys, err := m.cfg.githubServerKeys.GetKnownKeys()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, knownKey := range knownKeys {
+		if knownKey.Type() == key.Type() {
+			if bytes.Equal(knownKey.Marshal(), key.Marshal()) {
+				return nil
+			}
+		}
```

Comment:
This seems somewhat expensive and fragile; I wonder if we can do better with custom equality methods? If we know we are going to marshal the keys, should we cache the marshalled version of github keys?

Reply:
moved `key.Marshal` outside of the loop, though in reality there should be only one per type in the slice. I don't think marshaling is "very" expansive and likely called multiple times on each flow anyway.